crates/cranelift/src/gc/enabled.rs

```diff
@@ -56,7 +64,309 @@ pub fn gc_ref_table_fill_builtin(
+    let field_addr = func_env.prepare_gc_ref_access(builder, struct_ref, field_offset, field_size);
+
+    let field_val = match field_ty.element_type {
+        WasmStorageType::I8 => {
+            builder
+                .ins()
+                .load(ir::types::I8, ir::MemFlags::trusted(), field_addr, 0)
+        }
+        WasmStorageType::I16 => {
+            builder
+                .ins()
+                .load(ir::types::I16, ir::MemFlags::trusted(), field_addr, 0)
+        }
+        WasmStorageType::Val(_) => unreachable!(),
+    };
+
+    let extended = match extension {
+        Extension::Sign => builder.ins().sextend(ir::types::I32, field_val),
+        Extension::Zero => builder.ins().uextend(ir::types::I32, field_val),
+    };
+
+    Ok(extended)
+}
+
+pub fn translate_struct_get_s(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+    field_index: u32,
+    struct_ref: ir::Value,
+) -> WasmResult<ir::Value> {
+    translate_struct_get_and_extend(
+        func_env,
+        builder,
+        struct_type_index,
+        field_index,
+        struct_ref,
+        Extension::Sign,
+    )
+}
+
+pub fn translate_struct_get_u(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+    field_index: u32,
+    struct_ref: ir::Value,
+) -> WasmResult<ir::Value> {
+    translate_struct_get_and_extend(
+        func_env,
+        builder,
+        struct_type_index,
+        field_index,
+        struct_ref,
+        Extension::Zero,
+    )
+}
+
+pub fn translate_struct_set(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    struct_type_index: TypeIndex,
+    field_index: u32,
+    struct_ref: ir::Value,
+    new_val: ir::Value,
+) -> WasmResult<()> {
+    // TODO: See comment in `translate_struct_get` about the `trapz`.
+    builder.ins().trapz(struct_ref, ir::TrapCode::NullReference);
+
+    let field_index = usize::try_from(field_index).unwrap();
+    let interned_type_index = func_env.module.types[struct_type_index];
+
+    let struct_layout = func_env.struct_layout(interned_type_index);
+    let field_offset = struct_layout.fields[field_index];
+
+    let field_ty = match &func_env.types[interned_type_index].composite_type {
+        WasmCompositeType::Struct(s) => &s.fields[field_index],
+        _ => unreachable!(),
+    };
+
+    let field_size = wasmtime_environ::byte_size_of_wasm_ty_in_gc_heap(&field_ty.element_type);
+
+    // TODO: See comment in `translate_struct_get` about the `prepare_gc_ref_access`.
+    let field_addr = func_env.prepare_gc_ref_access(builder, struct_ref, field_offset, field_size);
+
+    match &field_ty.element_type {
+        WasmStorageType::I8 => {
+            builder
+                .ins()
+                .istore8(ir::MemFlags::trusted(), new_val, field_addr, 0);
```

Comment:
Is `::trusted()` the right default here? I would have thought that this would be "can possibly trap" flags. We can stuff in a debug trap code which doesn't cause trap table information to get emitted but it should still hinder optimizations and such in Cranelift (which I think we want at least for now?)

Reply:
I think we actually don't want to hinder optimizations. These aren't visible to Wasm semantics, or at least the whole point is they shouldn't be, so if the optimizer can prove that a load is redundant or whatever then I think we should let it.